pkg/ebpf/tracer.go

```diff
@@ -187,9 +220,310 @@ func (m *FlowTracer) Unregister() error {
+		Transport: readFlow.Transport}
+
+	aggRecord, collidedRecords, err := m.aggregateEntries(mapKey)
+	totalCollisions += len(collidedRecords)
+	if err == nil {
+		// Modify the readFlow record which was the evicted entry from the hash map
+		readFlow.Packets += aggRecord.Packets
+		readFlow.Bytes += aggRecord.Bytes
+		readFlow.FlowStartTime = aggRecord.FlowStartTime
+		readFlow.FlowEndTime = aggRecord.FlowEndTime
+		flowDuration := aggRecord.FlowEndTime - aggRecord.FlowStartTime
+		readFlow.TimeFlowStart = readFlow.TimeFlowEnd.Add(time.Duration(-flowDuration))
+	}
+	return err
+}
+
+func initResetMapEntries() {
+	resetMapValues = make([]recordValue, runtime.NumCPU())
+}
+
+// We Reset the entries before deletion, since we observe the entries are not cleared upon deletion,
+// and when a new flow maps to the same entry, it has some zombie entries.
+// TODO : Investigate if this is a generic issue and can it fixed in libbpf/gobpf?
+func (m *FlowTracer) resetEntriesAndDelete(mapKey recordKey) {
+	tlog := log.WithField("iface", m.interfaceName)
+	if mapKey.Transport.Protocol == 1 {
+		tlog.WithFields(logrus.Fields{
+			"mapKey": mapKey,
+		}).Debug("Deleting Record")
+	}
+	if mapKey.Direction == EGRESS {
+		if err := m.objects.XflowMetricMapEgress.Update(mapKey, resetMapValues, ebpf.UpdateExist); err != nil {
+			tlog.WithError(err).Warn("Failed in reset map")
+		}
+		if err := m.objects.XflowMetricMapEgress.Delete(mapKey); err != nil {
+			tlog.WithError(err).Warn("Failed in delete map")
+		}
+	} else {
+		if err := m.objects.XflowMetricMapIngress.Update(mapKey, resetMapValues, ebpf.UpdateExist); err != nil {
+			tlog.WithError(err).Warn("Failed in reset map")
+		}
+		if err := m.objects.XflowMetricMapIngress.Delete(mapKey); err != nil {
+			tlog.WithError(err).Warn("Failed in delete map")
+		}
+	}
+}
+
+// Monitor Egress Map to evict flows based on logic
+func (m *FlowTracer) MonitorEgress(ctx context.Context, evictionTimeout uint64, forwardFlows chan<- *flow.Record) {
+	tlog := log.WithField("iface", m.interfaceName)
+	go func() {
+		<-ctx.Done()
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			tlog.Debug("exiting monitor")
+			return
+		default:
+			egressMapIterator := m.objects.XflowMetricMapEgress.Iterate()
+			var mapEntries = 0
+			// Check Egress Map
+			var entriesAvail = true
+			timeNow := flow.Timestamp(monotime.Now())
+			for entriesAvail {
+				var mapKey recordKey
+				var mapValues []recordValue
+				entriesAvail = egressMapIterator.Next(&mapKey, &mapValues)
+				if !entriesAvail {
+					break
+				}
+				mapEntries++
+				aggRecord, collidedRecords := m.aggregateValues(mapKey, mapValues)
+
+				// Eviction Logic can be based on the following three metrics:
+				//   1) LastPacket > n sec ?
+				//   2) Packet/Byte count > Threshold ?
+				//   3) Number of Map entries > Threshold
+				// 			Perform more aggressive eviction by reducing the flow timeout duration
+				//   4) Regular Eviction (every n seconds)
```

Comment:
I have been digging into the code and maybe we don't need all the cases here, so the code would be simple and more stable.
E.g.:
- Can we delete case 1, considering that, in case 4 we will evict all flows every n seconds anyway.
- Can we delete case 2? I guess there is nothing wrong with a flow with many packets/bytes count.

So I'd evict as for the previous version of the agent:
1) When the number of map entries > CacheMaxFlows
2) Regularly, every CacheActiveTimeout

Reply:
Agree. I have just enumerated the options that could be implemented in future.